두번째 온보딩 리팩토링

theBettor opened the pull request:
## 🤷‍♂️ PR 내용(Issue 꼭 달기!)
1. 이슈가 많습니다. #121 
2. 민준님은 되는데 저는 카톡 로그인이 안되는 문제가 있어서 해결안되었거나 해야되는 부분은 추가로 이슈 만들어두었습니다. #130 
3. 다은님께 요청드린 부분의 답변이 아직안와서 기다리고 있는데 금방 해결되는 문제라 먼저 PR올립니다. #129

theBettor committed:
refactor: dependencies, lint, 화면 텍스트 수정, OnboardingLayout designSystem으로 편입

theBettor committed:
feat: 만들기, 입장 화면 모듈 추가

theBettor committed:
refactor: bottomBar를 선택적으로 추가할 수 있게 만듬

theBettor committed:
refactor:

DefaultButton이 CustomDropdownAndTextField의 selectedText와 name을 사용하려면 상위 Composable에서 상태 변수를 선언하고, CustomDropdownAndTextField와 DefaultButton에 전달해야 함.

selectedText = it 이런식으로 상태를 전달하고,
onSelectedTextChange: (String) -> Unit,     onNameChange: (String) -> Unit
를 만들어서                             onSelectedTextChange(option) 이런식으로 부모로 값을 전달한다.

✅ CustomDropdownAndTextField와 DefaultButton이 같은 상태를 공유

theBettor committed:
refactor : CustomDropdownAndTextField와 DefaultButton이 같은 상태를 공유

DefaultButton이 CustomDropdownAndTextField의 selectedText와 name을 사용하려면 상위 Composable에서 상태 변수를 선언하고, CustomDropdownAndTextField와 DefaultButton에 전달해야 함.

selectedText = it 이런식으로 상태를 전달하고,
onSelectedTextChange: (String) -> Unit,     onNameChange: (String) -> Unit
를 만들어서                             onSelectedTextChange(option) 이런식으로 부모로 값을 전달한다.

✅ CustomDropdownAndTextField와 DefaultButton이 같은 상태를 공유

theBettor committed:
Merge remote-tracking branch 'origin/refactor/onboarding_2' into refactor/onboarding_2

theBettor committed:
fix: 모듈 이름 헷갈려서 enter에서 make로 변경 #122

theBettor committed:
fix: enter에서 make로 변경2 #122

theBettor committed:
refactor: 온보딩모듈에 있던 그룹만들기 네비게이션 코드, 온보딩make모듈로 이동 그리고 화면연결 #122

theBettor committed:
fix: 클래스명 변경 #123

theBettor committed:
feat: RelationScreen으로 전환 #123

1. 릴레이션 스크린 완성이 안되었으니까 NavGraphBuilder는 완성을 안했었는데 이게 없으면 화면 전환이 불가능함
2. 하나의 Route 내에서 조건부로 화면을 변경하는 방법도 있지만 지금으로서는 하나의 Route 내에서 조건부로 화면을 변경하는 방법 이대로 유지하기로 했음.
3. 1번의 설명처럼 화면 완성이 안되었어도 NavGraphBuilder 완성해서 addOnboardingMakeRelationScreen()를 호출해줘야함.

theBettor committed:
feat: RelationScreen 완성 #123

theBettor committed:
feat : ProfileScreen으로 연결 및 파라미터명 변경 #124

theBettor committed:
fix: textfield 2개가 같이 동작해서 변수 하나 더 만들어서 상태관리 #124

theBettor committed:
feat: ProfileScreen만들고 InvitationScreen으로 전환 #125

theBettor committed:
feat: Enter모듈 구성 #127

theBettor committed:
feat: EnterSpaceScreen, HomeRoute로 보내기 #128

theBettor committed:
feat: 초대공유 화면 완성 #126

theBettor committed:
refactor: 화면 순서 변경 #129

theBettor committed:
refactor: SpeechBubble 사용하지 않고, Layout에서 BottomBar를 기준으로 위에 Image를 선택적으로 넣을 수 있게하기 #128

theBettor committed:
fix: 안쓰는 drawble 삭제, 화면연결 등 남은숙제 TODO로 기록

theBettor committed:
fix: 안쓰는 drawble 삭제, 화면연결 등 남은숙제 TODO로 기록 #128

theBettor committed:
Merge remote-tracking branch 'origin/refactor/onboarding_2' into refactor/onboarding_2

theBettor committed:
refactor: 버튼의 이미지, Box + clickable방식으로 변경

theBettor committed:
refactor: 버튼의 이미지, Box + clickable방식으로 변경 #131

theBettor committed:
Merge remote-tracking branch 'origin/refactor/onboarding_2' into refactor/onboarding_2

theBettor committed:
fix: 병합 중 화면연결 버그 수정

theBettor committed:
fix: enter모듈에 주석 때문에 lint에 걸림

신기한건 Format할 땐 안걸리는데 Check하면 걸리네..

agvber commented on core/designsystem/src/main/java/com/teampatch/core/designsystem/component/OnboardingLayout.kt:
```diff
@@ -0,0 +1,105 @@
+            }
+        }
+    ) { scaffoldPaddingValues ->
+        LazyColumn(
```
나중에 LazyColumn 없이 한번 Layout을 만들어 보면 좋을 것 같아요!

theBettor replied:
LazyColumn없이 라고 하면 어떤 형태일까요?

agvber replied:
```
) { scaffoldPaddingValues ->
          Column(
                    horizontalAlignment = Alignment.Start,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 16.dp)
                ) {
                    Text(
                        text = title,
                        fontFamily = PretendardFontFamily,
                        fontWeight = FontWeight.W700,
                        fontSize = 28.sp
                    )

                    Spacer(modifier = Modifier.height(16.dp))

                    Text(
                        text = subtext,
                        fontFamily = PretendardFontFamily,
                        fontWeight = FontWeight.W500,
                        fontSize = 18.sp
                    )
                }
                
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 34.dp)
                ) {
                    content()
                }
```

대충 이런느낌? 근데 Box도 없어도 될 것 같네요

theBettor committed:
feat: 두 모듈의 마지막 화면 HomeScreen으로 이동 #129

DJDrama commented on app/src/main/java/com/teampatch/harmony/MainNavHost.kt:
```diff
@@ -52,24 +58,55 @@ fun MainNavHost(
 
-        addOnboardingEnterSpaceScreen(onNextClick = { navController.navigateToHomeScreen() })
+        addOnboardingEnterSpaceScreen(
+            onBackRequest = navController::popBackStack,
```
한번 popBackStack이랑 navigateUp 확인해보시고 navigateUp을 사용해보시는 것도 고려해보세요~!

theBettor replied:
[18f70d9](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/18f70d9ad893443918feded9dc43b8296b17fa84)

DJDrama commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterInvitationCodeScreen.kt:
```diff
@@ -45,36 +35,47 @@ import androidx.compose.ui.text.withStyle
+    onBackRequest: () -> Unit,
+    onEnterSpaceScreenRequest: () -> Unit,
 ) {
     val code = remember { mutableStateListOf("", "", "", "", "") }
```
이 빈 문자열 5개는 뭘 의미하는거에요?

theBettor replied:
[4096797](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/40967973d318cb4e72cdc226a8c3ae0d8be53b46)

theBettor replied:
5개의 코드를 기억하기 위함이였습니다..!

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeProfileSettingsScreen.kt:
```diff
@@ -0,0 +1,137 @@
+                append(stringArrayResource(R.array.title_onboarding_make_profile)[1])
+            }
+            withStyle(style = SpanStyle(color = BL)) {
+                append(stringArrayResource(R.array.title_onboarding_make_profile)[2])
```
이런건 항상 index 0, 1, 2가 존재하는지 방어적으로 코딩해보면 좋을 것 같습니다! 왜냐하면 다른 개발자가 실수로 strings.xml에서 하나를 지워버리면 여기선 runtimeexception이 발생해서 좋지 않은 ux를 사용자에게 선사하니깐요!

theBettor replied:
[f705e18](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/f705e18024e3fdc35048ff662d718a4a7c4c55fc)

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeProfileSettingsScreen.kt:
```diff
@@ -0,0 +1,137 @@
+                Image(
+                    painter = rememberAsyncImagePainter(
+                        model = profileImageUri
+                            ?: com.teampatch.core.designsystem.R.drawable.ic_my_appbar, // 기본 이미지 설정
```
패키지명 지워버리세요~! ㅎㅎ

theBettor replied:
[4c226ce](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/4c226cee577738f3c6cd52fc92e53e4389742c5a)

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,164 @@
+                append(stringArrayResource(R.array.title_onboarding_make_relation)[1])
+            }
+            withStyle(style = SpanStyle(color = BL)) {
+                append(stringArrayResource(R.array.title_onboarding_make_relation)[2])
```
여기서도 배열에 인덱싱할땐 항상 주의하십쇼!

theBettor replied:
[4c226ce](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/4c226cee577738f3c6cd52fc92e53e4389742c5a)

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    ) {
+        Box(
+            modifier = Modifier.fillMaxWidth()
+        ) {
```
Column만 존재해도 되지않아요? Box가 필요할까요??

theBettor replied:
[cd4f8d6](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/cd4f8d6374082fd231aac43af7b5c8bb5f77edbf)

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,164 @@
+
+        Box(
+            modifier = Modifier.fillMaxWidth()
+        ) {
```
여기또한 단순히 Column(modifier = Modifier.fillMaxWidth()) 하면 되징 않을까요? Box를 없애고

theBettor replied:
[cd4f8d6](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/cd4f8d6374082fd231aac43af7b5c8bb5f77edbf)

DJDrama commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeViewModel.kt:
```diff
@@ -0,0 +1,33 @@
+        private set
+
+    var profileImageUri by mutableStateOf<Uri?>(null) // 프로필 이미지 상태 관리
+        private set
```
나중에 가능하다면 모두 MutableStateFlow로 변경하세요!
저도 민준님께 @agvber 잘못 말씀드린 부분이
MutableState을 사용해도 된다라고 했지만, 이렇게 되버리면 ViewModel 자체가 Jetpack Compose View에 완전하게 커플링이 되어버리기 때문에,
만약 Xml 뷰로 바꾸게 된다면 해당 뷰모델은 쓸모가 없어지니깐요.

하지만 StateFlow를 사용하면 XML이나 Compose에 모두 맞출 수 있는 뷰모델이 되니깐요!

theBettor replied:
[aa5929b](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/aa5929b39942717ab5f5c56728e25707c9f21531)

DJDrama commented on feature/onboarding/src/main/java/com/teampatch/feature/onboarding/login/ui/OnboardingFirstScreen.kt:
```diff
@@ -73,8 +73,8 @@ internal fun OnboardingFirstScreen(
-                .padding(top = 135.dp, bottom = 8.dp) // Column 시작 위치에 추가 패딩
+                .padding(paddingValues)
+                .padding(top = 135.dp, bottom = 8.dp)
         ) {
```
Column만으로 사용가능하다면 Box 모디파이어 설정들을 Column에 적용하고 Box는 없애도 될것같아요~!

theBettor replied:
[0183280](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/0183280738f06d0ad525570cab0cb7ef06f4bd1e)

theBettor committed:
refactor: popBackStack -> navigateUp

앱의 계층 구조를 따라 자연스럽게 상위화면(백스택)으로 이동하고 싶을 대 navigateUp, 특정화면 ID를 사용하여 특정화면으로 직접이동하고 싶을때 popBackStack

theBettor committed:
refactor:

1. stringArrayResource()는 @Composable 함수 내부에서만 호출 가능
→ OnBoardingLayout의 매개변수로 title: AnnotatedString을 받고 있기 때문에, 이 값을 넘겨주는 상위 @Composable에서 title을 생성하는 것이 적절합니다.

2. UI 재사용성을 유지
→ OnBoardingLayout 내부에서 titles를 선언하면 특정 문자열 배열에 의존하게 되어, 다른 화면에서도 OnBoardingLayout을 사용할 때 동일한 로직을 반복해야 합니다.
→ 반면, 상위 @Composable에서 title을 가공한 후 OnBoardingLayout에 전달하면, 더 유연하게 사용할 수 있습니다.

theBettor committed:
refactor: stringArrayResource()를 titles라는 변수를 만들어서 개발자 실수로 발생할 수 있는 runtimeException 방지

1. stringArrayResource()는 @Composable 함수 내부에서만 호출 가능
→ OnBoardingLayout의 매개변수로 title: AnnotatedString을 받고 있기 때문에, 이 값을 넘겨주는 상위 @Composable에서 title을 생성하는 것이 적절합니다.

2. UI 재사용성을 유지
→ OnBoardingLayout 내부에서 titles를 선언하면 특정 문자열 배열에 의존하게 되어, 다른 화면에서도 OnBoardingLayout을 사용할 때 동일한 로직을 반복해야 합니다.
→ 반면, 상위 @Composable에서 title을 가공한 후 OnBoardingLayout에 전달하면, 더 유연하게 사용할 수 있습니다.

theBettor committed:
fix: 패키지명삭제, 주석 삭제 등

theBettor committed:
fix: 패키지명삭제, 주석 삭제 등

theBettor committed:
refactor: Box제거하고 2개의 하위컬럼과 상위컬럼에서는 fillMaxWidth와 spaceBy로 여백주기

unfocusedBorderColor로 기본테두리 색상 지정, shape로 CornerShape 명확하게 지정

theBettor committed:
refactor: Box제거하고 2개의 하위컬럼과 상위컬럼에서는 fillMaxWidth와 spaceBy로 여백주기

unfocusedBorderColor로 기본테두리 색상 지정, shape로 CornerShape 명확하게 지정

theBettor committed:
refactor: box제거

theBettor committed:
refactor: box제거

theBettor committed:
fix: lint

theBettor committed:
fix: lint

theBettor committed:
1. 리스트 대신 MutableState와 String을 사용하면서 maxLength = 5로 길이제한 적용
2. isCodeComplete라는 변수로 5자리 입력 완료 여부 확인으로 bottomBar enable 컨트롤

theBettor committed:
refactor: Compose에 종속적인 ViewModel로 만드는 mutableStateOf이 아닌 XML에서도 활용할 수 있는 범용 ViewModel이 되는 MutableStateFlow 사용하기

1. mutableStateOf는 Compose의 State 관찰 시스템과 밀접하게 연관되어 있습니다.

Compose의 remember, snapshot system에서 자동으로 감지되고 recomposition을 트리거합니다.

하지만 XML 기반 View에서는 mutableStateOf를 직접 사용할 수 없습니다.

2. MutableStateFlow는 Kotlin Flow 기반으로 동작합니다.

XML 기반 View에서는 LiveData처럼 collect를 사용하여 구독할 수 있습니다.

Compose에서는 collectAsState()를 사용하여 관찰할 수 있습니다.

즉, Compose와 XML 둘 다 지원 가능하기 때문에 ViewModel의 UI 독립성이 증가합니다.

즉, ViewModel이 특정 UI 기술(Compose)에 의존하지 않도록 설계하는 것이 좋다는 의미입니다. 💡

theBettor committed:
Merge remote-tracking branch 'origin/refactor/onboarding_2' into refactor/onboarding_2

theBettor committed:
fix: lint

theBettor committed:
refactor: MutableStateFlow에 맞게 프로필이미지업데이트 변수 리팩토링

MutableStateFlow는 var이 아니라 val이므로 직접 할당할 수 없습니다.

대신 MutableStateFlow의 .value 속성을 사용하여 값을 업데이트합니다.

Compose에서는 collectAsState()를 사용하여 UI에서 관찰할 수 있습니다.

XML 기반 View에서는 flow.collect { }를 사용하여 변화를 감지할 수 있습니다.

agvber commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterInvitationCodeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+import com.teampatch.feature.onboarding.enter.R.string.subtext_onboarding_enter_invitation
+
+@Composable
+fun OnboardingEnterInvitationCodeScreen(
```
internal 처리 할 수 있으면 해주세요~

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterInvitationCodeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+    onEnterSpaceScreenRequest: () -> Unit,
+) {
+    val code = remember { mutableStateOf("") }
+    val maxLength = 5
```
상수는 위에 const val 형태로 지정해주세요~!

theBettor replied:
const val의 목적은 프로젝트 전역에서 사용하기 위함인거 같아서 companion object를 만들려고 하는데 지금은 또 Class가 아닌 internal 함수를 사용하니까 안되더라구요..

```kotlin
object Config {
    const val MAX_LENGTH = 5
}
```

이런식으로 object를 만들어도 되던데 굳이 이렇게 해야되나 싶긴한데.. 민준님의 생각은 어떠신가용? @agvber

agvber replied:
개념이 섞인 것 같아요! 
const val은 전역 변수를 말하는 것이 아니라 컴파일 시점에서 값을 할당한다고 생각하시면 됩니다
그러므로 값을 참조하는 경우에 오버헤드를 줄일 수 있으므로 요청드린겁니다

그리고 object로 사용하지 말고 코틀린이니까 아래와 같이 생성해도 동일하게 볼 수 있습니다.
```
private const val MAX_LENGTH = 5

@Composable
fun OnboardingEnterInvitationCodeScreen(
    onBackRequest: () -> Unit,
    onEnterSpaceScreenRequest: () -> Unit,
) {
    val code = remember { mutableStateOf("") }
```

theBettor replied:
[6c4af01](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/6c4af015627acbff0bc009b3653ec7bba356aac7) 프로젝트 전역이라고 말한게 틀린부분이였던거같군요

agvber commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterInvitationCodeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("다음")
```
String resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterInvitationCodeScreen.kt:
```diff
@@ -0,0 +1,157 @@
+    onBackRequest: () -> Unit,
+    onEnterSpaceScreenRequest: () -> Unit,
+) {
+    val code = remember { mutableStateOf("") }
```
by 같은 delegate 패턴을 한번 사용해보시면 좋을 것 같습니다!

theBettor replied:
[98a3732](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/98a3732db7f0c4b8ad026f650e419d63e8d61d55)

agvber commented on feature/onboarding-enter/src/main/java/com/teampatch/feature/onboarding/enter/OnboardingEnterSpaceScreen.kt:
```diff
@@ -0,0 +1,100 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("가족 공간 입장하기")
```
String resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeInviteGrandParentsScreen.kt:
```diff
@@ -0,0 +1,97 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("다음")
```
String resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeInviteGrandParentsScreen.kt:
```diff
@@ -0,0 +1,97 @@
+import com.teampatch.feature.onboarding.make.R.array.title_onboarding_make_invite_gp
+
+@Composable
+internal fun OnboardingMakeInviteGrandParentsScreen(
```
이거 초대코드 생성하는 화면이죠??
저라면 어짜피 할머니에게 초대하는 화면 보단 초대 페이지라는 것이 더 중요하니까
InviteCodeCreationScreen 이러한 형태로 만들 것 같긴하네요

@DJDrama 튜터님 생각은 어떠신가요!

theBettor replied:
저는 좋은거같아요!

theBettor replied:
[4d57468](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/4d57468e2fcaf81515211aa0e916f5ec5b313334)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("다음")
```
string resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    onShareInvitationScreenRequest: () -> Unit,
+) {
+    var selectedText by remember { mutableStateOf("") }
+    var name by remember { mutableStateOf("") } //
```
// <- 주석 지워주세요!

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    onShareInvitationScreenRequest: () -> Unit,
+) {
+    var selectedText by remember { mutableStateOf("") }
+    var name by remember { mutableStateOf("") } //
```
단순 remember 처리 하는 게 맞나요??
rememberSavable 사용해서 savedstatehandle에 저장해놓는 것이 좋지 않을까요?
온보딩 중간에 사용자가 다른 화면에 갔다오면 가끔 refresh 되는 경우가 있는데
그러한 상황을 대응하면 좋을 것 같아요!

theBettor replied:
[3b66985](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/3b66985a128d8ff382449e3aa398110ef97c16f5)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    onNameChange: (String) -> Unit,
+) {
+    var expanded by remember { mutableStateOf(false) }
+    val options = listOf("할머니", "할아버지")
```
string resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+    onNameChange: (String) -> Unit,
+) {
+    var expanded by remember { mutableStateOf(false) }
+    val options = listOf("할머니", "할아버지")
```
저라면 composable 함수 바깥에서 list 형태로 변수를 생성 할 것 같습니다! (or remember)
recomposition일어나면 계속해서 list 재할당하는 문제가 있을 것 같아요

theBettor replied:
지금 stringArrayResource를 사용하고 있는데, Composable calls are not allowed inside the calculation parameter of inline fun <T> remember(crossinline calculation: () -> TypeVariable(T)): TypeVariable(T)  이런 에러가 발생해서 어떻게 하면 좋을지 모르겠습니다.. @agvber

agvber replied:
stringArrayResource이면 그냥 패스하시져

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeParentsNameScreen.kt:
```diff
@@ -0,0 +1,164 @@
+            value = name,
+            onValueChange = { onNameChange(it) },
+            enabled = true,
+            placeholder = { Text("성함", color = Color.Gray) },
```
string resource 고곡씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeProfileSettingsScreen.kt:
```diff
@@ -0,0 +1,145 @@
+        }
+    )
+
+    val titles = stringArrayResource(title_onboarding_make_profile)
```
이거 remember로 감싸지 않아도 문제 없는 건가요??

@DJDrama

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeProfileSettingsScreen.kt:
```diff
@@ -0,0 +1,145 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("가족 공간 입장하기")
```
string resource 고고씽

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,162 @@
+                    .fillMaxWidth()
+                    .padding(20.dp)
+            ) {
+                Text("다음")
```
string resource ㄱㄱㄱ

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,162 @@
+                    append(titles[2])
+                }
+            } else {
+                append("Error: Missing Strings")
```
이건 뭔가ㅓ요??

theBettor replied:
UI 텍스트로 표시되게하는건데 생각해보니까 그럴 필요없어서 로그캣에 보이게 해야겠습니다.

theBettor replied:
[c55aadc](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/c55aadc72a2d5948e410edd89ab414f82d345268)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,162 @@
+                value = relation,
+                onValueChange = { onRelationChange(it) },
+                enabled = true,
+                placeholder = { Text("예) 손녀", color = Color.Gray) },
```
string resource ㄱㄱㄱ

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeRelationScreen.kt:
```diff
@@ -0,0 +1,162 @@
+                value = name,
+                onValueChange = { onNameChange(it) },
+                enabled = true,
+                placeholder = { Text("이름을 입력해 주세요.", color = Color.Gray) },
```
string resource ㄱㄱㄱ

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented on feature/onboarding-make/src/main/java/com/teampatch/feature/onboarding/make/OnboardingMakeViewModel.kt:
```diff
@@ -0,0 +1,36 @@
+import kotlinx.coroutines.flow.asStateFlow
+
+@HiltViewModel
+class OnboardingMakeViewModel @Inject constructor() : ViewModel() {
```
internal 처리하여 다른 모듈에서 접근 제한 할 수 있도록 해주세요!!

theBettor replied:
[2f55211](https://github.com/Team-PATCH/Harmony-Android/pull/135/commits/2f5521165a666d96365ca962fef24183ea1b6b60)

agvber commented:
히힣 제가 많이 까다롭네요 요즘 개발을 못하고 있어서 코드 보는게 재밌습니다
수정 완료 되면 재요청 부탁드려요 
글구 수고 하셨슴😊🙂😊🙂

theBettor committed:
fix: internal과 stringResource 추가

theBettor committed:
refactor: 특정 상수를 위한 object 생성

특정 클래스와 연관된 값이면 companion object 내부에 const val를 하고 싶어서 object를 만듬

theBettor committed:
refactor : by 키워드를 사용한 delegate 패턴으로 변경

🔍 왜 by를 쓰는 게 좋을까?
코드가 더 간결해짐

기존 코드 (mutableStateOf(""))에서는 .value를 계속 사용해야 함.

```kotlin
code.value = "12345"   // 기존 방식
println(code.value)    // 기존 방식
```
by 키워드를 사용하면 .value 없이 변수처럼 다룰 수 있음.

```kotlin
code = "12345"   // 개선된 방식
println(code)    // 개선된 방식
```
가독성이 향상됨

.value를 계속 사용하면 가독성이 떨어짐.

by를 쓰면 일반 변수처럼 사용할 수 있어 코드가 더 직관적임.

내부적으로 getValue 및 setValue를 자동으로 처리

by 키워드를 사용하면 getValue와 setValue를 자동으로 호출하므로 개발자가 직접 .value를 다룰 필요가 없음.

✅ 정리
방식	코드	장점
기존 (mutableStateOf)	val code = remember { mutableStateOf("") }	직접 .value를 사용해야 함
Delegate 패턴 (by)	var code by remember { mutableStateOf("") }	코드 간결, 가독성 향상, .value 생략 가능
🚀 즉, by를 사용하면 코드가 더 간결해지고 가독성이 좋아지므로 mutableStateOf를 사용할 때 권장됩니다! 🚀

theBettor committed:
refactor: 클래스명 변경

theBettor committed:
refactor: 중간에 사용자가 다른 화면에 갔다오면 가끔 refresh 되는 경우를 rememberSavable로 방지

✅ rememberSavable을 사용하면 해결 가능!
remember는 화면이 재구성될 때 상태를 유지하지만,
화면이 완전히 종료되었다가 다시 돌아오는 경우(예: 앱이 백그라운드로 갔다가 다시 실행됨) 상태가 사라질 수 있습니다.

이 문제를 방지하려면 rememberSavable을 사용하면 됩니다.

🛠 왜 rememberSavable을 사용해야 할까?
상태 저장 방식	화면 회전 시 유지	앱이 종료 후 다시 실행될 때 유지
remember	✅ 유지됨	❌ 사라짐
rememberSavable	✅ 유지됨	✅ 유지됨
remember는 구성 변경(Configuration Change, 예: 다크 모드 변경, 화면 회전) 에서는 유지되지만,
앱이 메모리에서 제거되었다가 복원될 때(예: 백그라운드에서 앱이 종료됨) 상태를 잃음.

rememberSavable은 내부적으로 SavedStateHandle이나 Bundle 을 사용하여
앱이 종료되었다가 다시 실행될 때도 데이터를 복원함.

🎯 추가로 SavedStateHandle을 활용하는 방법 (ViewModel 사용 시)
만약 ViewModel을 사용 중이라면, SavedStateHandle을 활용하여 상태를 저장할 수도 있습니다.

1️⃣ ViewModel에서 SavedStateHandle 사용
kotlin
복사
편집
class MyViewModel(savedStateHandle: SavedStateHandle) : ViewModel() {
    var selectedText by savedStateHandle.getLiveData("selectedText", "")
    var name by savedStateHandle.getLiveData("name", "")
}
2️⃣ Activity 또는 Composable에서 ViewModel 연결
```kotlin
val viewModel: MyViewModel = viewModel()
var selectedText by viewModel.selectedText
var name by viewModel.name
```
🚀 이렇게 하면 Jetpack Compose의 상태뿐만 아니라, 앱이 완전히 종료되었다가 다시 실행될 때도 데이터가 유지됩니다! 🚀

✅ 결론
빠르고 간단한 해결 방법 → rememberSavable 사용

ViewModel과 함께 사용 가능 → SavedStateHandle 활용

중요한 데이터라면 ViewModel을 통해 저장하는 것이 더 안전함

👉 만약 remember 때문에 상태가 사라지는 게 문제라면, rememberSavable로 교체하는 것이 가장 쉽고 효과적인 해결 방법입니다! 🚀

theBettor committed:
fix: append로 ui텍스트를 보여주는것이 아닌, 로그캣으로 확인하게하기

theBettor committed:
fix: const val로 변경

theBettor committed:
Merge branch 'develop' into refactor/onboarding_2
